### ISSUE TITLE
Reduce the complexity of approximate date fields by dropping some of ISO8601

### ISSUE THREAD

**Issue by stevenday** [openownership/data-standard#162]
Date fields that allow approximate values have a hugely complex regex listed in the schema docs for their validation. In writing BODS import code, I had to break this apart piece by piece to understand what kind of data I might be getting and I was a bit surprised by all the options it allowed.

In particular, I wonder whether we could drop the following:
- Times and timezones (all of the fields I could see are for *dates*, where a time wouldn't really add much useful data)
- Week numbers
- Ordinals (days of the year)
- Optional separators

This would leave it just allowing YYYY, YYYY-MM or YYYY-MM-DD (I think, I'm still not 100% sure if just a year is ok by the current regex). 

My rationale for this is that people are most likely going to be formatting their date/time data into ISO8601 anyway (from a local format, timestamp, etc), so I don't think the extra flexibility of these options is really helping them. On the contrary, by making the standard simpler and more explicit, I think it will help both publishers and consumers.

**Comment by ScatteredInk** [comment]
Agree with this - current regex is unhelpful and this would be a relatively easy fix.

**Comment by timgdavies** [comment]
See #12 for some background on dates. 

Research question for @stevenday and @kd-ods: Are we always seeing complete dates in incoming data? Or is handling incomplete / fuzzy dates showing up from experience as an important thing to consider?

**Comment by stevenday** [comment]
> Research question for @stevenday and @kd-ods: Are we always seeing complete dates in incoming data?

To the best of my knowledge all of the register's sources provide complete dates, so I'm minded to say it's not a very high priority at the moment.

**Comment by siwhitehouse** [comment]
We have a practical instance of the impact of including time in some fields. When we put example data through the Data Review Tool, it is appending 'T00:00:00+00:00' to the date in the following fields (at least):

- birthDate
- foundingDate
- startDate

As we are about to share this document with the overseas territories should we pick this off as a piece of work to do asap? If so, am I correct in thinking that we need to amend the regex? What else, if anything, needs to be done?

**Comment by Bjwebb** [comment]
> When we put example data through the Data Review Tool, it is appending 'T00:00:00+00:00' to the date in the following fields (at least):

I think this is only happening for spreadsheets. If so, its not quite being appended, but is an artifact of how we read the underlying date time information (different to how it is displayed) in flatten-tool. We can change what flatten-tool does, but it may have implications for exactly what change you want to make to the schema. Happy to have a brief chat about this.

**Comment by kd-ods** [comment]
For the record, the current date fields in the schema are listed below. To resolve this issue, we should decide whether a date or date-time is expected for each field and standardize how each is documented and validated in the schema. 

My first question is: we seem to have `"format": "date"` for some properties (e.g. `Annotation.creationDate`), but I don't think that's valid in JSON schema [Draft 4](https://datatracker.ietf.org/doc/html/draft-fge-json-schema-validation-00#section-7), is it? @rhiaro @odscjames - is that right, and is there any reason we should continue to use `"format": "date"`?

[Annotation.creationDate](https://standard.openownership.org/en/0.2.0/schema/reference.html#annotation)

[*Statement.statementDate](https://standard.openownership.org/en/0.2.0/schema/reference.html#entitystatement)

[EntityStatement.foundingDate](https://standard.openownership.org/en/0.2.0/schema/reference.html#entitystatement)

[EntityStatement.dissolutionDate](https://standard.openownership.org/en/0.2.0/schema/reference.html#entitystatement)

[Interest.startDate](https://standard.openownership.org/en/0.2.0/schema/reference.html#interest)

[Interest.endDate](https://standard.openownership.org/en/0.2.0/schema/reference.html#interest)

[PepStatusDetails.startDate](https://standard.openownership.org/en/0.2.0/schema/reference.html#pepstatusdetails)

[PepStatusDetails.endDate](https://standard.openownership.org/en/0.2.0/schema/reference.html#pepstatusdetails)

[PersonStatement.birthDate](https://standard.openownership.org/en/0.2.0/schema/reference.html#personstatement)

[PersonStatement.deathDate](https://standard.openownership.org/en/0.2.0/schema/reference.html#personstatement)

[PublicationDetails.publicationDate](https://standard.openownership.org/en/0.2.0/schema/reference.html#publicationdetails)

[Source.retrievedAt](https://standard.openownership.org/en/0.2.0/schema/reference.html#source)

**Comment by odscjames** [comment]
> My first question is: we seem to have "format": "date" for some properties (e.g. Annotation.creationDate), but I don't think that's valid in JSON schema Draft 4, is it?

No it is not. https://json-schema.org/draft-07/json-schema-release-notes.html#formats

But I don't think we can use "date" anyway. It's defined as "RFC 3339 “full-date”" - https://datatracker.ietf.org/doc/html/rfc3339#section-5.6 - ie Y AND M AND D. If we want to allow partials (Y or Y-M) then we can't use it.

So we are left with  `"type": "string"` and `"pattern":`.

> To resolve this issue, we should decide whether a date or date-time is expected for each field

I'd also add whether partial dates are allowed. Some fields there may be uncertanty and so it makes sense to allow them. Others like Annotation.creationDate or Source.retrievedAt I'm not sure allowing uncertainty is a good thing.

So go throught each field from the list above and answer:

* Is Y allowed?
* Is Y-M allowed?
* Is Y-M-D allowed?
* Is Y-M-D and a hours:minutes bit allowed?

If any hours:minutes bits are allowed, we can then return to the question of which time formats to allow, timezones, etc in a quest to achieve a simpler regex.

**Comment by kd-ods** [comment]
Thanks, @odscjames - really useful.

I've spun up a [sheet](https://docs.google.com/spreadsheets/d/1dQ1o0llbyxRSTCu0sX05AkGMB3RJEvzVmrJhswISv94/edit#gid=0) to do the work in (easier to convert into a docs table eventually). I'd be grateful for any thoughts on this.  (cc @StephenAbbott @siwhitehouse @rhiaro)

**Comment by lgs85** [comment]
Need to check how much of this can be resolved in #469

**Comment by kd-ods** [comment]
The work here is to:

- [ ] Finalise the spec of allowed date-times in [this table](https://docs.google.com/spreadsheets/d/1dQ1o0llbyxRSTCu0sX05AkGMB3RJEvzVmrJhswISv94/edit#gid=0). This should include a sense check against BO-related data sources that we've worked with (DK, SK, UK, GLEIF, FR.)
  - the spec aligns with what we see in DK data. Checked by @kd-ods  
- [ ] Establish how much of the revised spec can be implemented as part of the move to JSON Schema 2020-12 (#469)
- [ ] After the #469 is complete, open tickets for any remaining date-related issues.
